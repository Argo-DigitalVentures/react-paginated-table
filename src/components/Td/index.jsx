import React, { PropTypes } from 'react';
import get from 'lodash.get';

import {
  dataType,
  columnType,
  classNameType,
  styleType,
} from '../../propTypes';

import { evaluate } from '../../utils';

function Td(props) {
  const {
    column,
    columnIndex,
    columns,
    row,
    rowIndex,
    data,
    style: cellStyle,
    className: cellClassName,
    ...rest
  } = props;
  const value = (
    typeof column.selector === 'function' ?
    column.selector(row) :
    get(row, column.selector, null)
  );
  const dataProps = { value, column, columnIndex, columns, row, rowIndex, data };
  const className = evaluate(cellClassName, dataProps);
  const style = evaluate(cellStyle, dataProps);
  return (
    <td style={style} className={className} {...rest}>
      {
        typeof column.render === 'function' ? (
          <column.render {...dataProps} />
        ) : value
      }
    </td>
  );
}

Td.defaultProps = {
  style: {},
  className: '',
};

Td.propTypes = {
  data: dataType.isRequired,
  column: columnType.isRequired,
  columnIndex: PropTypes.number.isRequired,
  columns: PropTypes.arrayOf(columnType).isRequired,
  row: PropTypes.object.isRequired, // eslint-disable-line react/forbid-prop-types
  rowIndex: PropTypes.number.isRequired,
  style: styleType,
  className: classNameType,
};

export default Td;
