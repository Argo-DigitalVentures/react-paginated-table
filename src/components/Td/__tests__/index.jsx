import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Td from '../';

const data = [
  {
    company: 'Argo',
    location: 'New York',
    ceo: {
      firstName: 'Mark',
      lastName: 'Watson',
    },
  },
  {
    company: 'John Hancock',
    location: 'New York',
    ceo: {
      firstName: 'Andrew',
      lastName: 'Arnott',
    },
  },
];

const columns = [
  { id: 'company', header: 'Company Name', selector: 'company' },
  { id: 'location', header: 'Location', selector: 'location', render: () => null },
  { id: 'ceoFullName', header: 'CEO', selector: ({ ceo: { firstName, lastName } }) => `${firstName} ${lastName}` },
];

describe('<Td />', () => {
  it('should accept className, style, and other DOM props', () => {
    const tree = shallow(
      <Td
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        row={data[1]}
        rowIndex={1}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should accept className and style functions', () => {
    const styleFn = jest.fn().mockReturnValue({
      color: 'green',
    });
    const classNameFn = jest.fn().mockReturnValue('test');
    const tree = shallow(
      <Td
        style={styleFn}
        className={classNameFn}
        data={data}
        column={columns[0]}
        columnIndex={0}
        columns={columns}
        row={data[1]}
        rowIndex={1}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
    expect(styleFn).toBeCalledWith({
      value: 'John Hancock',
      column: columns[0],
      columnIndex: 0,
      columns,
      row: data[1],
      rowIndex: 1,
      data,
    });
    expect(classNameFn).toBeCalledWith({
      value: 'John Hancock',
      column: columns[0],
      columnIndex: 0,
      columns,
      row: data[1],
      rowIndex: 1,
      data,
    });
  });

  it('should use a custom renderer if available', () => {
    const tree = shallow(
      <Td
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        column={columns[2]}
        columnIndex={2}
        columns={columns}
        row={data[1]}
        rowIndex={1}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });

  it('should use a selector function to get the appropriate value from the row if available', () => {
    const tree = shallow(
      <Td
        style={{
          color: 'green',
        }}
        className="test"
        data={data}
        column={columns[1]}
        columnIndex={1}
        columns={columns}
        row={data[1]}
        rowIndex={1}
      />,
    );
    expect(toJson(tree)).toMatchSnapshot();
  });
});
