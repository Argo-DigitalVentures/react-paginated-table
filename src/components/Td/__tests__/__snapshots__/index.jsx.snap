exports[`<Td /> should accept className and style functions 1`] = `
<td
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  John Hancock
</td>
`;

exports[`<Td /> should accept className, style, and other DOM props 1`] = `
<td
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  John Hancock
</td>
`;

exports[`<Td /> should use a custom renderer if available 1`] = `
<td
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  Andrew Arnott
</td>
`;

exports[`<Td /> should use a selector function to get the appropriate value from the row if available 1`] = `
<td
  className="test"
  style={
    Object {
      "color": "green",
    }
  }>
  <render
    column={
      Object {
        "header": "Location",
        "id": "location",
        "render": [Function],
        "selector": "location",
      }
    }
    columnIndex={1}
    columns={
      Array [
        Object {
          "header": "Company Name",
          "id": "company",
          "selector": "company",
        },
        Object {
          "header": "Location",
          "id": "location",
          "render": [Function],
          "selector": "location",
        },
        Object {
          "header": "CEO",
          "id": "ceoFullName",
          "selector": [Function],
        },
      ]
    }
    data={
      Array [
        Object {
          "ceo": Object {
            "firstName": "Mark",
            "lastName": "Watson",
          },
          "company": "Argo",
          "location": "New York",
        },
        Object {
          "ceo": Object {
            "firstName": "Andrew",
            "lastName": "Arnott",
          },
          "company": "John Hancock",
          "location": "New York",
        },
      ]
    }
    row={
      Object {
        "ceo": Object {
          "firstName": "Andrew",
          "lastName": "Arnott",
        },
        "company": "John Hancock",
        "location": "New York",
      }
    }
    rowIndex={1}
    value="New York" />
</td>
`;
